File: php/12_sort/quicksort.php
<?php

function quickSort(array &$a)
{
    $n = count($a);

    quickSortInternally($a, 0, $n-1);
}

function quickSortInternally(array &$a, int $l, int $r)
{
    if ($l >= $r) return;

    $q = partition($a, $l, $r);
    quickSortInternally($a, $l, $q-1);
    quickSortInternally($a, $q+1, $r);
}

function partition(&$a, $l, $r): int
{
    $pivot = $a[$r];
    $i = $l;

    for ($j = $l; $j < $r; ++$j) {
        if ($a[$j] < $pivot) {
            [$a[$j], $a[$i]] = [$a[$i], $a[$j]];
            ++$i;
      }
    }

    [$a[$r], $a[$i]] = [$a[$i], $a[$r]];

    return $i;
}

$a1 = [1,4,6,2,3,5,4];
$a2 = [2, 2, 2, 2];
$a3 = [4, 3, 2, 1];
$a4 = [5, -1, 9, 3, 7, 8, 3, -2, 9];
quickSort($a1);
print_r($a1);
quickSort($a2);
print_r($a2);
quickSort($a3);
print_r($a3);
quickSort($a4);
print_r($a4);

File: scala/src/main/scala/ch12_sorts/QuickSort.scala
package ch12_sorts

object QuickSort {

  //find the K th smallest element int the array
  def findKthElement(items: Array[Int], k: Int): Int = {
    _findKthElement(items, k, 0, items.length - 1)
  }

  private[this] def _findKthElement(items: Array[Int], k: Int, p: Int, r: Int): Int = {
    val q = _partition(items, p, r)

    if (k == q + 1) {
      items(q)
    } else if (k < q + 1) {
      _findKthElement(items, k, p, q - 1)
    } else {
      _findKthElement(items, k, q + 1, r)
    }
  }

  def quickSort(items: Array[Int]): Array[Int] = {
    _quickSort(items, 0, items.length - 1)
    items
  }

  private[this] def _quickSort(items: Array[Int], p: Int, r: Int): Unit = {
    if (p >= r) {
      return
    }
    val q = _partition(items, p, r)
    _quickSort(items, p, q - 1)
    _quickSort(items, q + 1, r)
  }

  private[this] def _partition(items: Array[Int], p: Int, r: Int): Int = {
    val pivot = items(r)
    var i = p
    for (j <- Range(p, r)) {
      if (items(j) < pivot) {
        val temp = items(i)
        items(i) = items(j)
        items(j) = temp
        i += 1
      }
    }

    val temp = items(i)
    items(i) = items(r)
    items(r) = temp

    i
  }
}
